pkg/storage/snapshot/source.go

```diff
@@ -179,6 +184,71 @@ func (s *vmSnapshotSource) getVMRevision() (*snapshotv1.VirtualMachine, error) {
 	return vmRevision, nil
 }
 
+func (s *vmSnapshotSource) getControllerRevision(namespace, name string) (*appsv1.ControllerRevision, error) {
+	crKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := s.controller.CRInformer.GetStore().GetByKey(crKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to fetch ControllerRevision %s", crKey)
+	}
+	cr, ok := obj.(*appsv1.ControllerRevision)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected object format returned by informer")
+	}
+	return cr, nil
+}
+
+func (s *vmSnapshotSource) captureInstancetypeControllerRevision(namespace, name string) (string, error) {
+	existingCR, err := s.getControllerRevision(namespace, name)
+	if err != nil {
+		return "", err
+	}
+
+	snapshotCR := existingCR.DeepCopy()
+	snapshotCR.ObjectMeta.Reset()
+
+	// We strip out the source VM name from the CR name and replace it with the snapshot name
+	snapshotCR.Name = s.snapshot.Name + strings.ReplaceAll(existingCR.Name, s.snapshot.Spec.Source.Name, "")
```

Comment:
Repeat clones would generate unqiue VirtualMachineSnapshot names right?

Do you mean using the VirtualMachineSnapshot itself as a source instead of a VirtualMachine?

Reply:
Let's say we have a VM, named `"vm1"`, and a snapshot, named `"snapshot1"`. Now let's say we have a CR named `"snapshot1-vm1-randomString"`. In this case `s.snapshot.Spec.Source.Name` (which would be replaced by an empty string) is `vm1`.

So if we snapshot this VM again (via a new `"snapshot2"` object), we would get: `snapshot2-snapshot1-randomString`. So we stripped the VM name, but now we're left only with snapshot names.

If we do that again: `snapshot3-snapshot2-snapshot1-randomString`, and so on.

Am I completely off here?

Reply:
I think you're missing that when we restore we clone the CR again and rename it to be `vmName + randomString` where `randomString` is the instancetype/preference name, UID and generation. So `vm1-randomString` becomes `snapshot1-randomString` and is then restore to `newVM-randomString`. Additional snapshots of `vm1` would be `snapshot2-randomString` etc.

Reply:
-> `strings.Replace(existingCR.Name, s.snapshot.Spec.Source.Name, s.snapshot.Name, 1)`